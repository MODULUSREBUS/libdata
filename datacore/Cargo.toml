[package]
name = "datacore"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
index-access-storage = { git = "https://github.com/MODULUSREBUS/index-access" }
anyhow = "1.0.26"
futures-lite = "1.12.0"
byteorder = "1.3.4"
ed25519-dalek = "1.0.1"
blake3 = "1.3.1"
hex = "0.4"
rand = { version = "0.7.3", features = [ "std", "wasm-bindgen" ] }

[dev-dependencies]
index-access-memory = { git = "https://github.com/MODULUSREBUS/index-access" }
index-access-fs = { git = "https://github.com/MODULUSREBUS/index-access" }
insta = "1.8.0"
tokio = { version = "1.18.1", features = ["full"] }
crypto-hash = "0.3.4"
tempfile = "3.1.0"
criterion = { version = "0.3.4", features = [ "async_tokio" ] }
quickcheck = "0.9.2"
quickcheck_async = "0.1.1"
futures = "0.3.21"

[[bench]]
name = "io"
harness = false
