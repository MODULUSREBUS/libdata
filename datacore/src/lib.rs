#![forbid(unsafe_code, bad_style, nonstandard_style, future_incompatible)]
#![forbid(rust_2018_idioms, rust_2021_compatibility)]
#![deny(missing_debug_implementations)]
#![deny(missing_docs)]
#![cfg_attr(test, allow(dead_code))]
#![cfg_attr(test, deny(warnings))]

//! ## Introduction
//! Datacore is a secure, append-only, single-writer log.
//! It is meant for sharing large datasets and streams of real time data.
//! The primary way to use this crate is through the [Core] struct.
//!
//! ## Example
//! ```rust
//! # use futures_lite::future::FutureExt;
//!
//! use index_access_memory::IndexAccessMemory;
//! use datacore::{Core, generate_keypair};
//!
//! # #[tokio::main]
//! # async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
//! let keypair = generate_keypair();
//! let mut core = Core::new(
//!     IndexAccessMemory::new(),
//!     IndexAccessMemory::new(),
//!     IndexAccessMemory::new(),
//!     keypair.public, Some(keypair.secret)
//!     ).await?;
//!
//! core.append(b"hello", None).await?;
//! core.append(b"world", None).await?;
//!
//! assert_eq!(core.len(), 2);
//! assert_eq!(
//!     core.get(0).await?.unwrap().0,
//!     b"hello");
//! assert_eq!(
//!     core.get(1).await?.unwrap().0,
//!     b"world");
//! # Ok(())
//! # }
//! ```

mod block;
mod store_data;
mod store_blocks;
mod store_state;
mod merkle_tree_stream;
mod keys;
mod hash;
mod merkle;
mod core;

pub use index_access_storage::IndexAccess;
pub use block::{Signature, BlockSignature, Block, SIGNATURE_LENGTH};
pub use keys::{
    Keypair, PublicKey, SecretKey,
    generate_keypair, sign, verify
};
pub use hash::Hash;
pub use merkle::{Merkle, Node, NodeTrait};
pub use self::core::{Core, MAX_CORE_LENGTH, MAX_BLOCK_SIZE};
