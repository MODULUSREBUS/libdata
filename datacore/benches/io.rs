use criterion::{black_box, criterion_group, criterion_main, Criterion};
use tokio::runtime::Runtime;

use random_access_memory::RandomAccessMemory;
use datacore::{generate_keypair, Core};

type HomogenousCore<T> = Core<T, T, T>;
type MemoryCore = HomogenousCore<RandomAccessMemory>;

fn random_access_memory() -> RandomAccessMemory {
    RandomAccessMemory::new(1024)
}

async fn init() -> MemoryCore {
    let keypair = generate_keypair();
    Core::new(
        random_access_memory(),
        random_access_memory(),
        random_access_memory(),
        keypair.public, Some(keypair.secret))
        .await.unwrap()
}

async fn hypercore_append(mut core: MemoryCore, blocks: u64) {
    for i in 0..blocks {
        core.append(&i.to_be_bytes(), None).await.unwrap();
    }
}

pub fn criterion_benchmark(c: &mut Criterion) {
    c.bench_function("append 1000 blocks", |b| {
        let rt = Runtime::new().unwrap();
        b.to_async(rt).iter(|| async {
            let core = init().await;
            hypercore_append(black_box(core), black_box(1_000)).await;
        })
    });
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
