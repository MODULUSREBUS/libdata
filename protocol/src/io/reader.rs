use std::io::{Error, ErrorKind, Result};
use std::pin::Pin;
use std::time::Duration;
use std::task::{Context, Poll};
use std::future::Future;
use tokio::io::{AsyncRead, ReadBuf};
use futures_timer::Delay;

use crate::message::{Frame, FrameType};
use crate::noise::{Cipher, HandshakeResult};

const MAX_MESSAGE_SIZE: usize = crate::MAX_MESSAGE_SIZE as usize;

#[derive(Debug)]
enum Step {
    Header,
    Body { header_len: usize, body_len: usize },
}

#[derive(Debug)]
pub struct ReadState {
    /// The read buffer.
    buf: Vec<u8>,
    /// The end of the not-yet-processed byte range in the read buffer.
    end: usize,
    /// The logical state of the reading (either header or body).
    step: Step,
    /// The timeout after which the connection is closed.
    timeout: Option<Delay>,
    /// Timeout duration.
    timeout_duration: Option<Duration>,
    /// Optional encryption cipher.
    cipher: Option<Cipher>,
    /// The frame type to be passed to the decoder.
    frame_type: FrameType,
}

impl ReadState {
    pub fn new(timeout_ms: Option<u64>) -> Self {
        let timeout_duration = timeout_ms.map(Duration::from_millis);
        Self {
            buf: vec![0u8; MAX_MESSAGE_SIZE],
            end: 0,
            step: Step::Header,
            timeout: timeout_duration.map(Delay::new),
            timeout_duration,
            cipher: None,
            frame_type: FrameType::Raw,
        }
    }

    pub fn upgrade_with_handshake(&mut self, handshake: &HandshakeResult) {
        let mut cipher = Cipher::from_handshake_rx(handshake);
        cipher.apply(&mut self.buf[..self.end]);
        self.cipher = Some(cipher);
    }
    pub fn set_frame_type(&mut self, frame_type: FrameType) {
        self.frame_type = frame_type;
    }

    pub fn poll_reader<R>(
        &mut self,
        cx: &mut Context<'_>,
        mut reader: &mut R,
    ) -> Poll<Result<Frame>>
    where
        R: AsyncRead + Unpin,
    {
        loop {
            if let Some(result) = self.process() {
                return Poll::Ready(result);
            }

            let mut buf = ReadBuf::new(&mut self.buf[self.end..]);
            let n0 = buf.filled().len();
            let n = match Pin::new(&mut reader).poll_read(cx, &mut buf) {
                Poll::Ready(Ok(())) if (buf.filled().len() - n0) > 0 =>
                    buf.filled().len() - n0,
                Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
                // If the reader is pending, poll the timeout.
                Poll::Pending | Poll::Ready(Ok(_)) => {
                    // Return Pending if the timeout is pending, or an error if the
                    // timeout expired (i.e. returned Poll::Ready).
                    return match self.timeout.as_mut() {
                        None => Poll::Pending,
                        Some(mut timeout) => match Pin::new(&mut timeout).poll(cx) {
                            Poll::Pending => Poll::Pending,
                            Poll::Ready(_) => Poll::Ready(Err(
                                    Error::new(ErrorKind::TimedOut, "Remote timed out"))),
                        },
                    }
                }
            };

            let end = self.end + n;
            if let Some(ref mut cipher) = self.cipher {
                cipher.apply(&mut self.buf[self.end..end]);
            }
            self.end = end;

            // reset timeout
            match self.timeout_duration {
                None => None,
                Some(timeout_duration) =>
                    self.timeout.as_mut().map(|t| t.reset(timeout_duration)),
            };
        }
    }

    fn process(&mut self) -> Option<Result<Frame>> {
        if self.end == 0 {
            return None;
        }

        loop {
            match self.step {
                Step::Header => {
                    let mut body_len = 0;
                    let header_len = varinteger::decode(
                        &self.buf[..self.end],
                        &mut body_len,
                    );
                    let body_len = body_len as usize;
                    let message_len = header_len + body_len;

                    if message_len > MAX_MESSAGE_SIZE {
                        return Some(Err(Error::new(
                            ErrorKind::InvalidData,
                            "Message length above max allowed size",
                        )));
                    }
                    self.step = Step::Body {
                        header_len,
                        body_len,
                    };
                }
                Step::Body {
                    header_len,
                    body_len,
                } => {
                    let message_len = header_len + body_len;

                    if self.end < message_len {
                        return None;
                    }
                    else {
                        let frame = Frame::decode(
                            &self.buf[header_len..message_len],
                            &self.frame_type,
                        );
                        if self.end > message_len {
                            self.buf.copy_within(message_len..self.end, 0);
                        }
                        self.end -= message_len;
                        self.step = Step::Header;
                        return Some(frame);
                    }
                }
            }
        }
    }
}
