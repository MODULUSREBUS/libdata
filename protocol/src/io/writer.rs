use std::collections::VecDeque;
use std::fmt::Debug;
use std::io::Result;
use std::pin::Pin;
use std::task::{Context, Poll};
use tokio::io::AsyncWrite;

use crate::message::{EncodeError, Encoder, Frame};
use crate::noise::{Cipher, HandshakeResult};

const BUF_SIZE: usize = 1024 * 64;

#[derive(Debug)]
pub enum Step {
    Flushing,
    Writing,
    Processing,
}

#[derive(Debug)]
pub struct WriteState {
    queue: VecDeque<Frame>,
    buf: Vec<u8>,
    current_frame: Option<Frame>,
    start: usize,
    end: usize,
    cipher: Option<Cipher>,
    step: Step,
}
impl Default for WriteState {
    fn default() -> Self {
        Self {
            queue: VecDeque::new(),
            buf: vec![0u8; BUF_SIZE],
            current_frame: None,
            start: 0,
            end: 0,
            cipher: None,
            step: Step::Processing,
        }
    }
}
impl WriteState {
    pub fn upgrade_with_handshake(&mut self, handshake: &HandshakeResult) {
        self.cipher = Some(Cipher::from_handshake_tx(handshake));
    }

    pub fn queue_frame<F>(&mut self, frame: F)
    where
        F: Into<Frame>,
    {
        self.queue.push_back(frame.into())
    }

    fn try_queue_direct<T: Encoder>(
        &mut self,
        frame: &T,
    ) -> std::result::Result<bool, EncodeError> {
        let len = frame.encoded_len();
        if self.buf.len() < len {
            self.buf.resize(len, 0u8);
        }
        if len > self.remaining() {
            return Ok(false);
        }
        let len = frame.encode(&mut self.buf[self.end..])?;

        // advance
        let end = self.end + len;
        if let Some(ref mut cipher) = self.cipher {
            cipher.apply(&mut self.buf[self.end..end]);
        }
        self.end = end;

        Ok(true)
    }

    fn remaining(&self) -> usize {
        self.buf.len() - self.end
    }
    fn pending(&self) -> usize {
        self.end - self.start
    }

    pub fn poll_send<W>(&mut self, cx: &mut Context<'_>, mut writer: &mut W) -> Poll<Result<()>>
    where
        W: AsyncWrite + Unpin,
    {
        loop {
            self.step = match self.step {
                Step::Processing => {
                    if self.current_frame.is_none() && !self.queue.is_empty() {
                        self.current_frame = self.queue.pop_front();
                    }

                    if let Some(frame) = self.current_frame.take() {
                        if !self.try_queue_direct(&frame)? {
                            self.current_frame = Some(frame);
                        }
                    }
                    if self.pending() == 0 {
                        return Poll::Ready(Ok(()));
                    }
                    Step::Writing
                }
                Step::Writing => {
                    let n = match Pin::new(&mut writer)
                        .poll_write(cx, &self.buf[self.start..self.end])?
                    {
                        Poll::Ready(n) => n,
                        Poll::Pending => return Poll::Pending,
                    };
                    self.start += n;
                    if self.start == self.end {
                        self.start = 0;
                        self.end = 0;
                    }
                    Step::Flushing
                }
                Step::Flushing => {
                    match Pin::new(&mut writer).poll_flush(cx)? {
                        Poll::Ready(_) => (),
                        Poll::Pending => return Poll::Pending,
                    };
                    Step::Processing
                }
            }
        }
    }
}
