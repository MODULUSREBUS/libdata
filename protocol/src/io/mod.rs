mod reader;
mod writer;

use anyhow::{anyhow, Result};
use std::task::{Context, Poll};
use tokio::io::{AsyncRead, AsyncWrite};

use self::reader::ReadState;
use self::writer::WriteState;
use crate::message::{Frame, FrameType};
use crate::Options;
use crate::noise::HandshakeResult;

#[derive(Debug)]
pub struct IO<T> {
    io: T,
    options: Options,
    reader: ReadState,
    writer: WriteState,
}

impl<T> IO<T>
where
    T: AsyncRead + AsyncWrite + Send + Unpin + 'static,
{
    /// Create [IO].
    #[must_use]
    #[inline]
    pub fn new(io: T, options: Options) -> Self {
        let keepalive_ms = options.keepalive_ms;
        Self {
            io,
            options,
            reader: ReadState::new(keepalive_ms),
            writer: WriteState::default(),
        }
    }

    /// Check if `noise` is enabled.
    #[must_use]
    #[inline]
    pub fn noise_enabled(&self) -> bool {
        self.options.noise
    }
    /// Check if this is set as the initiator.
    #[must_use]
    #[inline]
    pub fn is_initiator(&self) -> bool {
        self.options.is_initiator
    }

    /// Upgrade with [HandshakeResult].
    #[inline]
    pub fn upgrade_with_handshake(&mut self, result: &Option<HandshakeResult>) {
        if self.options.encrypted {
            if let Some(handshake) = &result {
                self.reader.upgrade_with_handshake(handshake);
                self.writer.upgrade_with_handshake(handshake);
            }
        }
        self.reader.set_frame_type(FrameType::Message);
    }

    /// Poll for inbound messages and process them.
    #[inline]
    pub fn poll_inbound_read(&mut self, cx: &mut Context<'_>) -> Result<Option<Frame>> {
        let msg = self.reader.poll_reader(cx, &mut self.io);
        match msg {
            Poll::Ready(Ok(message)) => Ok(Some(message)),
            Poll::Ready(Err(e)) => Err(anyhow!(e)),
            Poll::Pending => Ok(None),
        }
    }

    /// Poll for outbound messages and write them.
    #[inline]
    pub fn poll_outbound_write(&mut self, cx: &mut Context<'_>) -> Result<()> {
        if let Poll::Ready(Err(e)) = self.writer.poll_send(cx, &mut self.io) {
            Err(anyhow!(e))
        } else {
            Ok(())
        }
    }

    /// Queue a frame to be sent.
    #[inline]
    pub fn queue_frame<F: Into<Frame>>(&mut self, frame: F) {
        self.writer.queue_frame(frame);
    }
}
