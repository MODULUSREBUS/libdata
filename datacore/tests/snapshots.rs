mod common;
use common::storage_fs;

use std::path::Path;
use std::fs::File;
use std::io::Read;
use tokio::test;
use tempfile;
use insta;

use datacore::{Core, Keypair};

fn read_bytes(dir: &Path, s: &str) -> Vec<u8> {
    let mut f = File::open(dir.join(s)).unwrap();
    let mut bytes = Vec::new();
    f.read_to_end(&mut bytes).unwrap();
    bytes
}

const KEYPAIR_BYTES: [u8; 64] = [
    86, 29, 202, 51, 72, 159, 192, 155, 76, 131, 249, 122, 241, 244, 9, 244,
    157, 139, 190, 59, 130, 201, 224, 241, 121, 161, 171, 30, 158, 108, 23, 0,
    184, 16, 141, 118, 116, 37, 127, 146, 105, 139, 107, 124, 101, 123, 166,
    152, 83, 209, 170, 236, 172, 23, 111, 253, 30, 197, 241, 83, 169, 233,
    237, 77
];

#[test]
pub async fn snapshots_append() {
    let dir = tempfile::tempdir().unwrap().into_path();
    let keypair = Keypair::from_bytes(&KEYPAIR_BYTES).unwrap();
    let mut core = Core::new(
        storage_fs(&dir.to_path_buf().join("data")).await,
        storage_fs(&dir.to_path_buf().join("blocks")).await,
        storage_fs(&dir.to_path_buf().join("merkle")).await,
        keypair.public, Some(keypair.secret))
        .await.unwrap();

    let data = b"abcdef";
    for &b in data {
        core.append(&[b], None).await.unwrap();
    }
    assert_eq!(core.len() as usize, data.len());

    let mut blocks_store = Vec::new();
    let mut data_store = Vec::new();
    for (i, _) in data.iter().enumerate() {
        let b = read_bytes(&dir, &format!("blocks/{}", i));
        let d = read_bytes(&dir, &format!("data/{}", i));
        blocks_store.extend(b);
        data_store.extend(d);
    }
    let state_store = read_bytes(&dir, "merkle/state");

    insta::assert_debug_snapshot!(blocks_store);
    insta::assert_debug_snapshot!(data_store);
    insta::assert_debug_snapshot!(state_store);
}
