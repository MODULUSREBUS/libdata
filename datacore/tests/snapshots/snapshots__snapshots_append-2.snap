---
source: datacore/tests/snapshots.rs
expression: data_store

---
[
    97,
    98,
    99,
    100,
    101,
    102,
]
